### PR TITLE
Rule based editor improvements

### PR DESCRIPTION
**Pull request by jouvin**
Read pending changes (in https://github.com/quattor/configuration-modules-grid/pull/89) forgot before moving the rule-based editor to CAF. Includes:
- Enhanced handling of hashes: possibility to write all keyword/value pairs on the same line
- Support for option subset in rule conditions
- Fixes

### CODE REVIEW

**Comment by jouvin** [comment]
Relevant changes from https://github.com/quattor/configuration-modules-grid/pull/89 merged into CAF::RuleBasedEditor (and its tests). Ready for merging.

File: src/main/perl/RuleBasedEditor.pm [hunk 1]
```diff
@@ -148,27 +152,27 @@ end of the line if it is modified by L<CAF::RuleBasedEditor>.
 
 =item *
 
-LINE_FORMAT_KEY_VAL:        key val (e.g. Xrootd, Apache)
+LINE_FORMAT_KW_VAL:        key val (e.g. Xrootd, Apache)
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: what does `KW` stand for? (also, `KV` as abbreviation for key-value is also used elsewhere)

File: src/main/perl/RuleBasedEditor.pm [hunk 2]
```diff
@@ -193,11 +197,11 @@ LINE_VALUE_ARRAY: the value is an array. Rendering controlled by LINE_VALUE_OPT_
 
 =item
 
-LINE_VALUE_HASH_KEYS: the value is hash whose keys are the value. Rendering similar to arrays.
+LINE_VALUE_HASH: the value is a hash of string. Rendering controlled by LINE_VALUE_OPT_xxx constants.
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: `string`-> `strings`

File: src/main/perl/RuleBasedEditor.pm [hunk 3]
```diff
@@ -193,11 +197,11 @@ LINE_VALUE_ARRAY: the value is an array. Rendering controlled by LINE_VALUE_OPT_
...
 
 =item
 
-LINE_VALUE_STRING_HASH: the value is a hash of string. Rendering controlled by LINE_VALUE_OPT_xxx constants.
+LINE_VALUE_HASH_KEYS: the value is hash whose keys are the value. Rendering similar to arrays.
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: `hash`  -> `a hash`
similat to arrays, you mean `similar to LINE_VALUE_ARRAY`?

File: src/main/perl/RuleBasedEditor.pm [hunk 4]
```diff
@@ -437,10 +454,31 @@ sub _formatAttributeValue
...
 
+    } elsif ( $value_fmt == LINE_VALUE_HASH && !($value_opt & LINE_VALUE_OPT_SINGLE) ) {
+        $self->debug(2, "$function_name: hash received with keys: ", join(",",(sort keys %$attr_value)));
+        # Key/value separator (D: space)
+        my $key_val_separator = ' ';
+        $key_val_separator = ':' if ($value_opt & LINE_VALUE_OPT_SEP_COLON);
+        # Prefix to add before key (D: none)
+        my $key_prefix = '';
+        $key_prefix = '-' if ($value_opt & LINE_KEY_OPT_PREFIX_DASH);
+        $self->debug(2,"$function_name: key prefix: >>>$key_prefix<<<; key/value separator: $key_val_separator");
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: space after comma

File: src/main/perl/RuleBasedEditor.pm [hunk 5]
```diff
@@ -437,10 +454,31 @@ sub _formatAttributeValue
...
+        my @tmp_values;
+        foreach my $k (sort keys %$attr_value) {
+            my $v = $attr_value->{$k};
+            # Keys may be escaped if they contain characters like '/': unescaping a non-escaped
+            # string is generally harmless.
+            my $tmp = $key_prefix.unescape($k).$key_val_separator.$v;
+            $self->debug(2,"$function_name: hash key/value string: '".$tmp);
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: space after comma, there probably needs to be a `."'" after the `$tmp` (but you use `$tmp` directly in the double quoted string)

File: src/main/perl/RuleBasedEditor.pm [hunk 6]
```diff
@@ -437,10 +454,31 @@ sub _formatAttributeValue
...
+            push @tmp_values, $tmp;
+        }
+        $formatted_value = join " ", @tmp_values;
+
     } elsif ($value_fmt == LINE_VALUE_HASH_KEYS) {
         $formatted_value = join " ", sort keys %$attr_value;
 
-    } elsif (($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_STRING_HASH)) {
+    } elsif ( ($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_HASH) || ($value_fmt == LINE_VALUE_ARRAY) ) {
+        # Do nothing when LINE_VALUE_HASH or LINE_VALUE_ARRAY and LINE_VALUE_OPT_SINGLE
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: comment does not match logic above, also no `LINE_VALUE_OPT_SINGLE` is used

File: src/main/perl/RuleBasedEditor.pm [hunk 7]
```diff
@@ -568,13 +606,20 @@ sub _buildLinePattern
...
+        # config_value: scape characters with a specific meaning in the regexp
         $config_value =~ s/\\/\\\\/g;
-        $config_value =~ s/([\-\+\?\.\*\[\]()\^\$])/\\$1/g;
+        $config_value =~ s/([\-\+\?\.\*\[\]()\^\$\{\}])/\\$1/g;
         $config_value =~ s/\s+/\\s+/g;
     } else {
         $config_value = "";
     }
 
+    # config_param: escape characters with a specific meaning in the regexp.
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: if this is the same escapiong as above, put it in a (anonymous?) sub instead of copy pasting

File: src/main/perl/RuleBasedEditor.pm [hunk 8]
```diff
@@ -838,7 +885,10 @@ sub _parse_rule
         $condition = '';
     }
 
-    # Check if rule condition is met if one is defined
+    # Check if rule condition is met if one is defined.
+    # A condition consists on the existence of a given option set (called condition set)and optionally
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: space after `)`

File: src/main/perl/RuleBasedEditor.pm [hunk 9]
```diff
@@ -857,22 +907,25 @@ sub _parse_rule
         $self->debug(2, "$function_name: condition option set = '$cond_option_set', ",
                      "condition attribute = '$cond_attribute', negate=$negate");
         my $cond_satisfied = 1;    # Assume condition is satisfied
+        my $cond_true = 1;
+        my $tmp_options = $config_options;
+        foreach my $set (split /\//, $cond_option_set) {
+          if ( $tmp_options->{$set} ) {
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: 4 space indent

File: src/main/perl/RuleBasedEditor.pm [hunk 10]
```diff
@@ -857,22 +907,25 @@ sub _parse_rule
         $self->debug(2, "$function_name: condition option set = '$cond_option_set', ",
                      "condition attribute = '$cond_attribute', negate=$negate");
         my $cond_satisfied = 1;    # Assume condition is satisfied
+        my $cond_true = 1;
+        my $tmp_options = $config_options;
+        foreach my $set (split /\//, $cond_option_set) {
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: can you put `()` around the split arguments? it looks a but odd otherwise

File: src/main/perl/RuleBasedEditor.pm [hunk 11]
```diff
@@ -1104,28 +1158,12 @@ sub _apply_rules
...
-                                                       $line_fmt,
-                                                       $value_fmt,
-                                                       $value_opt,
-                                                      );
-                        $self->_updateConfigLine($keyword, $config_value, $line_fmt, 1);
-                    }
-                    $config_updated = 1;
+          
+                } elsif ( $value_fmt == LINE_VALUE_HASH ) {
+                    # Hash are not processed immediately. First, all the values from all the options sets
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: `Hash is` or `Hash values are`

File: src/main/perl/RuleBasedEditor.pm [hunk 12]
```diff
@@ -1164,15 +1202,43 @@ sub _apply_rules
...
+            # pairs are concatenated to create the value.
+            if ( $value_opt & LINE_VALUE_OPT_SINGLE ) {
+                foreach my $k (sort keys %hash_values) {
+                    my $v = $hash_values{$k};
+                    # Value is made by joining key and value as a string
+                    # Keys may be escaped if they contain characters like '/': unescaping a non-escaped
+                    # string is generally harmless.
+                    my $tmp = unescape($k)." $v";
+                    $self->debug(1,"$function_name: formatting (string hash) attribute '".$rule_info->{attribute}."' value ($tmp, value_fmt=$value_fmt)");
+                    $config_value = $self->_formatAttributeValue($tmp,
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: handle undef return values

File: src/main/perl/RuleBasedEditor.pm [hunk 13]
```diff
@@ -1164,15 +1202,43 @@ sub _apply_rules
...
+                                                                 $line_fmt,
+                                                                 $value_fmt,
+                                                                 $value_opt,
+                                                                );
+                    $self->_updateConfigLine($keyword,$config_value,$line_fmt,1);
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: space after comma

Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: also handle undef

File: src/test/perl/rbe_rule_parser.t [hunk 14]
```diff
@@ -350,7 +406,7 @@ set_file_contents($DPM_CONF_FILE,$DPM_INITIAL_CONF_1);
 
 # Test  simple variable substitution
 set_file_contents($DPM_CONF_FILE,$DPM_INITIAL_CONF_1);
```
Comment (stdweird) [inline src/test/perl/rbe_rule_parser.t]: put all this in a function instead of copy/pasting, like
```perl
sub test1 
{
    my ($fn, $data, $args,$expected, $msg);
    set_file_contents($fn,$data);
    my $fh = CAF::RuleBasedEditor->open($fn, log => $obj);
    ok(defined($fh), "$fn was opened $msg");
     $changes = $fh->updateFile(@$args);
    is("$fh", $expected, "$fn has expected contents $msg");
    $fh->close();
}
test1($DPM_CONF_FILE, $DPM_INITIAL_CONF_1, [\%dpm_config_rules_1, $dpm_options], $DPM_EXPECTED_CONF_1,"(config 1)");
```

File: src/test/perl/rbe_value_format.t [hunk 15]
```diff
@@ -140,35 +140,64 @@ Readonly my @TEST_ARRAY => ('confFile', 'logFile', 'unused', 'logKeep', 'logFile
 Readonly my $FORMATTED_ARRAY => 'confFile logFile unused logKeep logFile';
 Readonly my $FORMATTED_ARRAY_SORTED => 'confFile logFile logFile logKeep unused';
 Readonly my $FORMATTED_ARRAY_UNIQUE => 'confFile logFile logKeep unused';
-my $rbe_fh = CAF::RuleBasedEditor->open($FILENAME, log => $obj);
+# When LINE_VALUE_OPT_SINGLE is set, formatAttributeValue() does nothihng and returns the array
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: typo in nothing

File: src/test/perl/rbe_value_format.t [hunk 16]
```diff
@@ -140,35 +140,64 @@ Readonly my @TEST_ARRAY => ('confFile', 'logFile', 'unused', 'logKeep', 'logFile
...
+Readonly my $FORMATTED_ARRAY_SINGLE => 'ARRAY\(0x[a-f\d]+\)';
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: how is this a "feature" that is unittested? should return undef because it tried to convert impossible type

File: src/test/perl/rbe_value_format.t [hunk 17]
```diff
@@ -140,35 +140,64 @@ Readonly my @TEST_ARRAY => ('confFile', 'logFile', 'unused', 'logKeep', 'logFile
...
                                                   LINE_VALUE_ARRAY,
                                                   LINE_VALUE_OPT_UNIQUE,
                                                  );
 is($formatted_value, $FORMATTED_ARRAY_UNIQUE, "Array values (unique) correctly formatted");
+$formatted_value = $rbe_fh->_formatAttributeValue(\@TEST_ARRAY,
+                                                  LINE_FORMAT_KW_VAL,
+                                                  LINE_VALUE_ARRAY,
+                                                  LINE_VALUE_OPT_SINGLE,
+                                                 );
+like($formatted_value, qr/$FORMATTED_ARRAY_SINGLE/, "Array values (single) correctly formatted");
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: if this actually returns an arrayref, use `is_deeply($formatted_value, [qw(a b c)], "returns arrayref")` to test the retruned arrayref (and not e.g. a string like the regex you are testing with)

File: src/test/perl/rbe_value_format.t [hunk 18]
```diff
@@ -140,35 +140,64 @@ Readonly my @TEST_ARRAY => ('confFile', 'logFile', 'unused', 'logKeep', 'logFile
...
+                                                  LINE_VALUE_HASH,
+                                                  0,
+                                                 );
+is($formatted_value, $FORMATTED_STRING_HASH, "String hash correctly formatted");
+$formatted_value = $rbe_fh->_formatAttributeValue(\%STRING_HASH,
+                                                  LINE_FORMAT_KW_VAL,
+                                                  LINE_VALUE_HASH,
+                                                  LINE_VALUE_OPT_SINGLE,
+                                                 );
+like($formatted_value, qr/$FORMATTED_STRING_HASH_SINGLE/, "String hash (1 key/val per line) correctly formatted");
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: use `is_deeply` to test if it returns the hashref you expect

File: src/main/perl/RuleBasedEditor.pm (hunk 1 above)
Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: I found `KEY` misleading and had in mind that `KW` as `keyworkd` was probably better. I'm happy to revert it or change to something else if you don't agree...

**Comment by jouvin** [comment]
@stdweird as usual, thank you for your careful/detailed review! I'll fix things soon...

File: src/main/perl/RuleBasedEditor.pm (hunk 1 above)
Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: I checked but didn't see `KV` anywhere. Would you prefer `LINE_FORMAT_PARAM_VALUE` instead?

Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: apologies, with elsewhere i meant outside of quattor. i don't mind the exact name that much, but after the 16.4 release it should change anymore (or changes would require backwards compatible code).

Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: Exactly this is why I try to clean it up! And I am open to any suggestion better than the current one! The change is pretty easy to do at this point, just a matter of a couple of sed commands.

File: src/main/perl/RuleBasedEditor.pm (hunk 7 above)
Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: Done.

File: src/main/perl/RuleBasedEditor.pm (hunk 12 above)
Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: Done. Error handling significantly improved!

File: src/test/perl/rbe_value_format.t (hunk 17 above)
Comment (jouvin) [inline src/test/perl/rbe_value_format.t]: I don't understand. `_formatAttributeValue()` always returns a string. In this case, this is a multiline string.

File: src/test/perl/rbe_value_format.t (hunk 18 above)
Comment (jouvin) [inline src/test/perl/rbe_value_format.t]: Same remark as before. `_formatAttributeValue')` really returns a string.

File: src/main/perl/RuleBasedEditor.pm [hunk 19]
```diff
@@ -437,10 +456,32 @@ sub _formatAttributeValue
...
+            push @tmp_values, $tmp;
+        }
+        $formatted_value = join " ", @tmp_values;
+
     } elsif ($value_fmt == LINE_VALUE_HASH_KEYS) {
         $formatted_value = join " ", sort keys %$attr_value;
 
-    } elsif (($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_STRING_HASH)) {
+    } elsif ( ($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_HASH) || ($value_fmt == LINE_VALUE_ARRAY) ) {
+        # In addition to LINE_VALUE_AS_IS, do nothing when ether LINE_VALUE_HASH or LINE_VALUE_ARRAY and
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: ether -> either

File: src/main/perl/RuleBasedEditor.pm [hunk 20]
```diff
@@ -532,12 +573,36 @@ sub _formatConfigLine
...
+
+Arguments:
+    regexp_str: initial regexp string (characters not escaped)
+
+Return value:
+    string: regexp with all specail characters escaped
+
+=cut
+
+sub _escape_regexp_string {
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: `{` on new line

Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: would also be nice that this metohd itself is unittested, not only when called from within `_buildLinePattern`

File: src/main/perl/RuleBasedEditor.pm [hunk 21]
```diff
@@ -629,19 +699,19 @@ sub _commentConfigLine
 
     unless ($config_param) {
         $self->error("$function_name: 'config_param' argument missing (internal error)");
-        return 1;
+        return undef;
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: just `return;` is also ok.

File: src/main/perl/RuleBasedEditor.pm [hunk 22]
```diff
@@ -1074,12 +1160,16 @@ sub _apply_rules
...
                 # no longer part of the configuration in a still existing LINE_VALUE_ARRAY or
-                # LINE_VALUE_STRING_HASH.
+                # LINE_VALUE_HASH.
                 unless ($attribute_present) {
                     if ($rule_parsing_options->{remove_if_undef}) {
                         $self->debug(1, "$function_name: attribute '$rule_info->{attribute}' undefined, ",
                                      "removing configuration line");
-                        $self->_commentConfigLine($keyword, $line_fmt);
+                        unless ( $self->_commentConfigLine($keyword, $line_fmt) ) {
+                            $self->error("$function_name: _commentConfigLine() encountered an internal error, ",
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: if `_commentConfigLine` already reports an error, no need to repeat that here i think

File: src/main/perl/RuleBasedEditor.pm (hunk 12 above)
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: yes it has. thanks for that!

File: src/main/perl/RuleBasedEditor.pm [hunk 23]
```diff
@@ -1164,15 +1244,55 @@ sub _apply_rules
...
+                                                                 $value_fmt,
+                                                                 $value_opt,
+                                                                );
+                     unless ( defined($config_value) ) {
+                        $self->error("$function_name: _formatAttributeValue() encountered an internal error, ",
+                                     "rule $rule_id ignored.");
+                        $ruleset_status = 0;
+                        next;
+                     }
+                    $self->_updateConfigLine($keyword, $config_value, $line_fmt, 1);
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: fix indentation

File: src/test/perl/rbe_value_format.t (hunk 17 above)
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: the comment above `$FORMATTED_ARRAY_SINGLE` says otherwise. also the pattern itself looks like a stringified arrayref, instead of a newline joined list of strings.

File: src/test/perl/rbe_value_format.t (hunk 18 above)
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: same remark about comment above the pattern and the pattern itself being a stringified hashref

**Comment by jouvin** [comment]
As for me really ready for merging!!! No other modification/improvement planned for this release!

File: src/main/perl/RuleBasedEditor.pm [hunk 24]
```diff
@@ -482,7 +482,13 @@ sub _formatAttributeValue
     } elsif ( ($value_fmt == LINE_VALUE_AS_IS) || ($value_fmt == LINE_VALUE_HASH) || ($value_fmt == LINE_VALUE_ARRAY) ) {
         # In addition to LINE_VALUE_AS_IS, do nothing when either LINE_VALUE_HASH or LINE_VALUE_ARRAY and 
         # LINE_VALUE_OPT_SINGLE (if it is not set, this is processed before so no need to test it again). 
-        $formatted_value = $attr_value;
+        # Just ensure that the value can be interpolated as a string (it is a scalar).
+        if ( ref(\$attr_value) eq 'SCALAR' ) {
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: since you already know `attr_value` is defined, you can do `if(ref($attr_value) eq '') {` for scalars

File: src/main/perl/RuleBasedEditor.pm [hunk 25]
```diff
@@ -482,7 +482,13 @@ sub _formatAttributeValue
...
+            $formatted_value = $attr_value;
+        } else {
+            $self->error("$function_name: value '$attr_value' cannot be interpolated as a string");
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: i would add all (or at least more) of the original arguments to the message, otherwise it will not be very usefull when debugging

File: src/test/perl/rbe_value_format.t [hunk 26]
```diff
@@ -167,7 +165,9 @@ $formatted_value = $rbe_fh->_formatAttributeValue(\@TEST_ARRAY,
                                                   LINE_VALUE_ARRAY,
                                                   LINE_VALUE_OPT_SINGLE,
                                                  );
-like($formatted_value, qr/$FORMATTED_ARRAY_SINGLE/, "Array values (single) correctly formatted");
+# When LINE_VALUE_OPT_SINGLE is set, formatAttributeValue() should return undef 
+# if the value cannot be interpolated as a string
+ok(!$formatted_value, "An array passed as value cannot be interpolated as a string");
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: should be `ok(! defined($formatted_value), ` to for undef

File: src/test/perl/rbe_value_format.t [hunk 27]
```diff
@@ -195,9 +194,17 @@ is($formatted_value, $FORMATTED_STRING_HASH, "String hash correctly formatted");
...
+is($formatted_value, $FORMATTED_STRING_HASH_2, "String hash correctly formatted (key prefixed with ')");
+$formatted_value = $rbe_fh->_formatAttributeValue(\%STRING_HASH,
+                                                  LINE_FORMAT_KW_VAL,
+                                                  LINE_VALUE_HASH,
                                                   LINE_VALUE_OPT_SINGLE,
                                                  );
-like($formatted_value, qr/$FORMATTED_STRING_HASH_SINGLE/, "String hash (1 key/val per line) correctly formatted");
+# When LINE_VALUE_OPT_SINGLE is set, formatAttributeValue() should return undef 
+# if the value cannot be interpolated as a string
+ok(!$formatted_value, "A hash passed as value cannot be interpolated as a string");
```
Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: same remark for undef

File: src/main/perl/RuleBasedEditor.pm [hunk 28]
```diff
@@ -223,7 +231,7 @@ use enum qw(
 
 =pod
 
-=head3 LINE_VALUE_OPT_xxx: options for rendering the value
+=head3 LINE__VALUE_OPT_xxx: options for rendering the config line
```
Comment (stdweird) [inline src/main/perl/RuleBasedEditor.pm]: one `_`?

Comment (jouvin) [inline src/main/perl/RuleBasedEditor.pm]: It has been fixed during the rename of these options, It seems your comment doesn't apply to last version in the PR...

File: src/test/perl/rbe_value_format.t (hunk 27 above)
Comment (jouvin) [inline src/test/perl/rbe_value_format.t]: I don't understand this one. It duplicates the previous one, is not it?

Comment (stdweird) [inline src/test/perl/rbe_value_format.t]: the only way to test if formatAttributeValue returns undef (and not eg empty string or 0), is to use `ok(! defined(),`

Comment (jouvin) [inline src/test/perl/rbe_value_format.t]: Sure, I understood it but it was what your previous comment was saying, isn't it?

**Comment by jouvin** [comment]
I added a few additional unit tests to cover better a few corner case identified in previous mods. I think I address all comments...! @stdweird anything else ?!

**Comment by stdweird** [comment]
nope. LGTM
